### DISCUSSION TITLE
feature: Animation Manager

### DISCUSSION THREAD

**Discussion by WeylonSantana** [AscensionGameDev/Intersect-Engine#1475] [Ideas]
Sit down and get your coffee, this is a discussion for developers (not that I'm one), I appreciate if you leave your opinion so we can reach a logical and quick conclusion.

**Is your feature request related to a problem? Please describe.**
Currently the animations are defined and coded internally. We have _attack, _cast, _idle, _shoot, _weapon and that doesn't bring the developers a dynamic interaction between the developer and the players experience.

"But how Weylon? Explain yourself better!"

So, it is normal that we have different types of weapons in our games, but only 1 animation defined for that type of weapon. Considering the case of the _shoot animation, if we have a bow and arrow we can make this animation, but this animation will be the same for shooting shuriken, shooting crossbows, throwing knives, because these other types of weapons require different animations, and we have only 1 _shoot animation, which is not dynamic. The same goes for _weapon for example, different types of weapons require different types of animation.

So we got to the point that animation management needs to become dynamic

**Describe the solution you'd like**
I'm going to present a solution (which may or may not be crazy, because I'm not a developer), and you developer reading this feel free to refute me, throw my idea in the trash, or improve it, it doesn't matter, I'd just like it we came to a conclusion that possibly some well-known programmer would do it (since I have no knowledge so I won't do it)

My idea is that the animations were defined in the server's config, through an array. And for each weapon's attack animations to be set in the item editor.
In the item editor there is already "attack animation" that runs an animation in front of the attack. This for ages has been a simple solution for the intersect, but today it is no longer necessary due to _attack or _weapon, but which is still useful and can complement these aforementioned animations by adding more effect and sound.

An array on the server - In the same way that we can put items into categories, or something similar, we would define the suffixes that sprite images should have on the server. For example:

"SpriteAnimations" : [
   "lance",
   "sword",
   "bow",
   "crossbow",
   ...
]

So with some magic (programming obviously), we would define in the item editor, which animation it will use, and then with more magic (more fucking programming) the client would get the base sprite that is being used by the class editor and would try to search for the sprite with the selected suffix to render the animation at the time of the attack.

In idiotic terms: we create the animations on the server, we define it in the editor, so on the client it will search for "player_sword.png", "player_bow.png" etc.

Could I be being an idiot and talking nonsense? Yes, but I'm being a pretty logical idiot, aren't I?

Well, in my view this would make the animations dynamic, each weapon could have its own animation (or for the laziest like me, use the same animation, but not stick to it).

Logically there should be a lot of bug prevention if the image doesn't exist, etc. And also configuration (like the ones that currently exist) for the amount of frames.

You can have a dictionary approach to this: Dictionary<Animation, Number of Frames>
"SpriteAnimations" : [
   "lance" => 4
   "sword" => 2
   "bow" => 3
   "crossbow" => 4
   ...
]

**Additional context**
But anyway, I started the discussion, and I hope there can be more contributions to a really viable solution (I don't think mine really is), it was just a kick-off to make games made with the intersect visually better in their quality, and more fun for the player experience.

I'd like to say that I think it's fair that only the walking animation should be hardcoded.
About the cast, it should follow the same model as above, because different spells might need different casts, but I don't know how it would be implemented.
And about idle, we could have more than one, but that would be too luxury haha

**Comment by oomstdev** [comment]
I know this is a bit old, but I'm on here looking for something and saw this. 

Really can't animations be anything? Like _attack could be a sleeping animation for all the game knows. Would be kind of silly to have an attack button put someone to the sleeping animations, but anyhow. Wouldn't it be a matter of making it _attack1 _attack2, etc. Then it's up to  the game designer to decide what is 1 or 2. You could even have the option for the weapons to use which animation, 1, 2, etc.

**Reply by WeylonSantana** [reply]
@oomstdev I appreciate your comment, so I implemented this function with the help of Panda. Was Added to the Class Editor, Item Editor and Spell Editor, a new combobox, allowing users to select their own custom "_animations", for attack(unarmed and armed) and cast. 

About pngs file, must follow this format "spriteName_animation_customAnimation.png.png"

Examples:
base-male.png
base-male_attack.png ==> Default sprite attack animation
base-male_attack_sword.png ==> Custom animation
base-male_cast.png ==> Default sprite cast animation
base-male_cast_kamehameha.png ==> Custom animation
and so on...

More details on - https://github.com/AscensionGameDev/Intersect-Engine/pull/1347

It's not an animation manager, but it already allows more animations.

**Comment by pandinocoder** [comment]
Weylon showed me this video today and I actually think it's a great basis for an animation manager that I would want us to have:

[![Alt text](https://img.youtube.com/vi/KAPJLY1JWlA/0.jpg)](https://www.youtube.com/watch?v=KAPJLY1JWlA)

It definitely is not perfect, but it has some pretty neat (and reasonable) ideas.

**Comment by da-rulez** [comment]
For me personally, an editor would be perfect. Instead of spritesheets you have frames (or one big sheet with all the frames, whatever works best performance wise). 

In the editor you pick the frames you want, in what order and how much time you want between frames. -> I think it would be best to have your frames be 4 directional. So if you pick one frame, it's actually one column with 4 frames (for the 4 cardinal directions).

Save that as a sprite animation and in item / spell editor you can just assign the sprite animation you want. + This would be cool if you can have an event command: Play Sprite animation X
